### PR TITLE
Added option to customize partition events by stage logic subclasses

### PR DESCRIPTION
**Pull request by IgorFedchenko**
This is a small PR, which adds a way to customize partition events handling by different stage logic classes.

Before this PR partitions events were handled by `SingleSourceStageLogic` without any option to override it. Besides, original kafka driver callbacks also provide `consumer` instance passed to callbacks, so actually handling could use some consumer's methods.

Now consumer is passed to this callbacks, and the callbacks itself may be customized by logic subclasses (looks like only for `TransactionalSourceLogic` so far).

### CODE REVIEW

**Comment by IgorFedchenko** [comment]
During PR #48 review process, I thought that `IPartitionEventHandler` interface will be made public. But currently all references in original alpakka project, that use this interface, are private/internal. So actually making it public will not give library users any option of use this interface - so keeping it internal so far. We can switch it to public whenever there will be any option to use it from outside.

File: src/Akka.Streams.Kafka/Helpers/PartitionEventHandler.cs [hunk 1]
```diff
@@ -7,53 +7,59 @@
...
+    /// Allows to execute user code when Kafka rebalances partitions between consumers, or an Alpakka Kafka consumer is stopped.
+    /// Use with care: These callbacks are called synchronously on the same thread Kafka's `poll()` is called.
+    /// A warning will be logged if a callback takes longer than the configured `partition-handler-warning`.
+    ///
+    /// There is no point in calling `CommittableOffset`'s commit methods as their committing won't be executed as long as any of
+    /// the callbacks in this class are called.
     /// </summary>
     [ApiMayChange]
-    internal interface IPartitionEventHandler
+    internal interface IPartitionEventHandler<K, V>
```
Comment (Aaronontheweb) [inline src/Akka.Streams.Kafka/Helpers/PartitionEventHandler.cs]: LGTM

File: src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs [hunk 2]
```diff
@@ -61,7 +61,9 @@ protected override IActorRef CreateConsumerActor()
             var partitionsAssignedHandler = GetAsyncCallback<IEnumerable<TopicPartition>>(PartitionsAssigned);
             var partitionsRevokedHandler = GetAsyncCallback<IEnumerable<TopicPartitionOffset>>(PartitionsRevoked);
 
-            var eventHandler = new AsyncCallbacksPartitionEventHandler(partitionsAssignedHandler, partitionsRevokedHandler);
+            IPartitionEventHandler<K, V> eventHandler = new AsyncCallbacksPartitionEventHandler<K,V>(partitionsAssignedHandler, partitionsRevokedHandler);
+            // This allows to override partition events handling by subclasses
+            eventHandler = AddToPartitionAssignmentHandler(eventHandler);
```
Comment (Aaronontheweb) [inline src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SingleSourceStageLogic.cs]: LGTM

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs [hunk 3]
```diff
@@ -499,15 +499,20 @@ public override void OnPartitionsAssigned(IImmutableSet<TopicPartition> partitio
...
+                
                 _actor._rebalanceInProgress = false;
             }
 
             public override void OnPartitionsRevoked(IImmutableSet<TopicPartitionOffset> partitions)
             {
-                // TODO: Add warning if IPartinionEventHandler will be public and call takes more then _warningDuration
-                _partitionEventHandler.OnRevoke(partitions);
+                var watch = Stopwatch.StartNew();
+                _partitionEventHandler.OnRevoke(partitions, _restrictedConsumer);
```
Comment (Aaronontheweb) [inline src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs]: LGTM

File: src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs [hunk 4]
```diff
@@ -499,15 +499,20 @@ public override void OnPartitionsAssigned(IImmutableSet<TopicPartition> partitio
...
+                CheckDuration(watch, "onRevoke");
+
```
Comment (Aaronontheweb) [inline src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs]: @IgorFedchenko do we need to `.Dispoe()` `Stopwatch` here?

Comment (IgorFedchenko) [reply src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs]: It does not implement `IDisposable` interface, but I do `watch.Stop()` in `CheckDuration` method implementation. 
But while technically this is correct, this looks unobvious - I thought about that when was writing this, but decided to save coding lines :) Added explicit `Stop()` call before passing `watch` to `CheckDuration` method, now should look better.

Comment (Aaronontheweb) [reply src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs]: Ah got it, thanks for explaining

**Comment by Aaronontheweb** [comment]
@IgorFedchenko FYI, on the previous commit had one test failure:

Akka.Streams.Kafka.Tests.Akka.Streams.Kafka.Tests.Integration.PlainSourceIntegrationTests.PlainSource_consumes_messages_from_KafkaProducer_with_subscribe_to_topic

You can see the logs and such if you click through here: https://github.com/akkadotnet/Akka.Streams.Kafka/runs/227053848

Might have a race condition of some kind with that test.

**Comment by IgorFedchenko** [comment]
@Aaronontheweb Created [issue](https://github.com/akkadotnet/Akka.Streams.Kafka/issues/57) for that - looks little bit wierd, so unfortunately can not fix this right now, "on the fly".

**Pull request merged**